compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs

```diff
@@ -1433,6 +1435,84 @@ fn build_vtable_type_di_node<'ll, 'tcx>(
     .di_node
 }
 
+fn vcall_visibility_metadata<'ll, 'tcx>(
+    cx: &CodegenCx<'ll, 'tcx>,
+    ty: Ty<'tcx>,
+    trait_ref: Option<PolyExistentialTraitRef<'tcx>>,
+    vtable: &'ll Value,
+) {
+    enum VCallVisibility {
+        Public = 0,
+        LinkageUnit = 1,
+        TranslationUnit = 2,
+    }
+
+    let Some(trait_ref) = trait_ref else { return };
+
+    let trait_ref_self = trait_ref.with_self_ty(cx.tcx, ty);
+    let trait_ref_self = cx.tcx.erase_regions(trait_ref_self);
+    let trait_def_id = trait_ref_self.def_id();
+    let trait_vis = cx.tcx.visibility(trait_def_id);
```

Comment:
I think that the closest existing concept similar  to `vcall_visibility` would be reachability, rather than visibility.

<details><summary>Example with an incorrectly optimized out vtables</summary>

```rust
trait Foo { fn foo(&self) { println!("foo") } }
trait Bar { fn bar(&self) { println!("bar") } }
trait Baz { fn baz(&self) { println!("baz") } }

impl Foo for usize {}
impl Bar for usize {}
impl Baz for usize {}

pub struct FooBox(Box<dyn Foo>);
pub struct BarBox(Box<dyn Bar>);
pub struct BazBox(Box<dyn Baz>);

pub fn make_foo() -> FooBox { FooBox(Box::new(0)) }
pub fn make_bar() -> BarBox { BarBox(Box::new(0)) }
pub fn make_baz() -> BazBox { BazBox(Box::new(0)) }

#[inline]
pub fn f(a: FooBox) { a.0.foo() }
pub fn g<T>(b: BarBox) { b.0.bar() }
#[inline]
fn h(c: BazBox) { c.0.baz() }
pub const H: fn(BazBox) = h;
```

</details>

Reply:
Thanks for pointing this out. I don't really have a solution for this right now. I tried looking at the reachability of the `trait_ref` and the members of the traits with the [`privacy_access_levels`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.privacy_access_levels) query. But for all of the traits and the functions in the traits (queried with `tcx.vtable_entries`) this just returns `None`, meaning it is not reachable IIUC.

I guess I would somehow have to check the reachability for the `dyn Trait` types. But I couldn't figure out how to do that.

Do you have ideas on how to best address this issue?